Introduce 64-bit unsigned long field type

mayya-sharipova opened the pull request:
This field type supports
- indexing of integer values from [0, 18446744073709551615]
- precise queries (term, range) 
- precise sorting and terms aggs. 
- other aggs are based on double values.

Sort values, term aggs keys, script values, doc values fields, 
source fields return long (for values <=2^63-1)
 and BigInteger (for values > 2^63-1).

Closes #32434

mayya-sharipova committed:
Introduce 64-bit unsigned long field type

This field type supports
- indexing of integer values from [0, 18446744073709551615]
- precise queries (term, range)
- sorting and aggregations is based on conversion of long values
  to double and can be imprecise for large values.

Closes #32434

elasticmachine commented:
Pinging @elastic/es-search (:Search/Mapping)

jimczi reviewed (commented):
> sorting and aggregations is based on conversion of long values
to double and can be imprecise for large values.

I did not check the pr yet but I wonder why sorting cannot be done on the unsigned value, We can use `Long#compareUnsigned` in a custom comparator ?

mayya-sharipova commented:
@jimczi Thanks, Jim, I will look into this.

jpountz commented on docs/reference/mapping/types/numeric.asciidoc:
```diff
@@ -115,7 +116,7 @@ The following parameters are accepted by numeric types:
 
     Try to convert strings to numbers and truncate fractions for integers.
-    Accepts `true` (default) and `false`.
+    Accepts `true` (default) and `false`. Not applicable for unsigned_long.
```
```suggestion
    Accepts `true` (default) and `false`. Not applicable for `unsigned_long`.
```

jpountz commented on docs/reference/mapping/types/unsigned_long.asciidoc:
```diff
@@ -0,0 +1,147 @@
+integer with a minimum value of 0 and a maximum value of +2^64^-1+
+(from 0 to 18446744073709551615).
+
+At index-time, an indexed value is converted to the singed long range:
```
```suggestion
At index-time, an indexed value is converted to the signed long range:
```

jpountz commented on docs/reference/mapping/types/unsigned_long.asciidoc:
```diff
@@ -0,0 +1,147 @@
+At index-time, an indexed value is converted to the singed long range:
+[- 9223372036854775808, 9223372036854775807] by subtracting +2^63^+ from it
+and stored as a singed long taking 8 bytes.
+At query-time, the same conversion is done on query terms.
```
I'm unsure whether this implementation detail is worth sharing.

jpountz commented on docs/reference/mapping/types/unsigned_long.asciidoc:
```diff
@@ -0,0 +1,147 @@
+
+Unsigned long is a numeric field type that represents an unsigned 64-bit
+integer with a minimum value of 0 and a maximum value of +2^64^-1+
+(from 0 to 18446744073709551615).
```
In Elasticsearch range queries with the from/to notation, the "to" bound is exclusive, while I think it's inclusive in your example here. Maybe be explicit by saying "to XYZ included" ?

jpountz commented on docs/reference/mapping/types/unsigned_long.asciidoc:
```diff
@@ -0,0 +1,147 @@
+--------------------------------
+//TEST[continued]
+
+Range queries can contain ranges with decimal parts.
```
Clarify what happens in that case?

jpountz commented on libs/x-content/src/main/java/org/elasticsearch/common/xcontent/json/JsonXContentParser.java:
```diff
@@ -166,6 +167,26 @@ public long doLongValue() throws IOException {
+        } else { // for all other value types including numbers with decimal parts
+            throw new IllegalArgumentException("For input string: [" + parser.getValueAsString() + "].");
+        }
+    }
```
do we really need to have support for unsigned longs at such a low level, or could we implement parsing of unsigned longs in the field mapper directly?

jpountz commented on x-pack/plugin/mapper-unsigned-long/src/main/java/org/elasticsearch/xpack/unsignedlong/UnsignedLongFieldMapper.java:
```diff
@@ -0,0 +1,519 @@
+     * @param value – unsigned long value in the range [0; 2^64-1], values greater than 2^63-1 are negative
+     * @return signed long value in the range [-2^63; 2^63-1]
+     */
+    private static long convertToSignedLong(long value) {
```
without context I'd guess that this is doing a cast, maybe make the method name more explicit, e.g. `unsignedToSortableSignedLong`

jpountz reviewed (commented):
This looks like a great start. I agree with Jim's comment that it would be great if sorting was accurate, like ranges.

mayya-sharipova committed:
Address feedback

mayya-sharipova committed:
Merge remote-tracking branch 'upstream/master' into unsigned64bits_integer

mayya-sharipova committed:
Modifications after master merge

mayya-sharipova committed:
Merge remote-tracking branch 'upstream/master' into unsigned64bits_integer

mayya-sharipova commented:
@jimczi  @jpountz Thanks for the initial review.  I have addressed your comments, please continue to review when you have time.

mayya-sharipova committed:
Rename methods

jimczi commented on docs/reference/mapping/types/unsigned_long.asciidoc:
```diff
@@ -0,0 +1,156 @@
+Search queries across several numeric types one of which `unsigned_long` are supported,
+except queries with sort. Thus, a sort query across two indexes where the same field
+is `unsigned_long` in one index, and `long` in another, doesn't produce correct results
+and must be avoided. If there is a need for a such kind of sorting, script based
```
```suggestion
and must be avoided. If there is a need for such kind of sorting, script based
```

jimczi commented on docs/reference/mapping/types/unsigned_long.asciidoc:
```diff
@@ -0,0 +1,156 @@
+and must be avoided. If there is a need for a such kind of sorting, script based
+sorting can be used instead.
+Aggregations across several numeric types one of which `unsigned_long` are supported,
+except a terms aggregation.
```
We handle long => double conversion in the `terms` aggregation so I wonder if that's true ?

mayya-sharipova replied:
@jimczi Indeed,  I have modified the code in 9e057c0 to support `terms` agg on mixed numeric types with `unsigned_long`

jimczi replied:
++, thanks

jimczi commented on server/src/main/java/org/elasticsearch/search/DocValueFormat.java:
```diff
@@ -48,6 +49,8 @@
 /** A formatter for values as returned by the fielddata/doc-values APIs. */
 public interface DocValueFormat extends NamedWriteable {
+    long MASK_2_63 = 0x8000000000000000L;
+    BigInteger BIGINTEGER_2_64_MINUS_ONE = BigInteger.ONE.shiftLeft(64).subtract(BigInteger.ONE); // 2^64 -1
```
These variables should be final and static ?

mayya-sharipova replied:
@jimczi Thanks Jim, I get a hint from Intellij that `final static` declarations are unnecessary here, as interface variables are by default `final` and `static`.

jimczi replied:
oh that's an interface, right. Sorry for the noise

jimczi commented on server/src/main/java/org/elasticsearch/common/io/stream/StreamOutput.java:
```diff
@@ -826,7 +827,14 @@ public final void writeOptionalInstant(@Nullable Instant instant) throws IOExcep
+                (o, v) -> {
+                    o.writeByte((byte) 26);
+                    o.writeString(v.toString());
+                }
```
Can you fix the indentation of the blocks ?

mayya-sharipova replied:
Addressed in 9e057c0

jimczi commented on x-pack/plugin/mapper-unsigned-long/src/main/java/org/elasticsearch/xpack/unsignedlong/UnsignedLongScriptDocValues.java:
```diff
@@ -0,0 +1,67 @@
+        if (in.advanceExact(docId)) {
+            resize(in.docValueCount());
+            for (int i = 0; i < count; i++) {
+                values[i] = (Number) DocValueFormat.UNSIGNED_LONG_SHIFTED.format(in.nextValue());
```
We'd need to provide the doc value format but that could be done in the `ScriptDocValues.Long` directly ?

mayya-sharipova replied:
@jimczi  Thanks for the feedback. I am not very clear what you are suggesting here. Are you suggesting to use `ScriptDocValues.Longs` instead of creating a new class?   If yes, I have opted to create a new class, as `ScriptDocValues.Longs` hold values in a array  `long[] values`, while we need an array of `Number[] values` as sometimes produced values can be of type `long` and other times of type `BigInteger`.

jimczi replied:
So today `ScriptDocValues.Long` returns a `Long` but it could return a `Number` instead and rely on a provided doc values format to do the conversion to a number. In the normal case we would conver to a `Long` but if the provided doc values format is the `unsigned_long` then it can also return a `BigInteger` in some cases. 
I am not sure it's a good idea so feel free to keep the code separated if you prefer. In any case I don't think we need to rely on a `Number[]` array. We can store the unsigned long in a plain `long[]` array and do the conversion to `Number` lazily when `getValue` or `get(int)` are called.

mayya-sharipova replied:
@jimczi  Thanks Jim, this is a good idea. I explored using `Longs` but it turned out it involves more modifications as `Longs::getValue` expects `long`.  Instead, I followed your 2nd idea to used `long[]` for values in `UnsignedLongScriptDocValues.java `  and this is addressed in 2b567c96.

jimczi replied:
++, thanks

jimczi reviewed (approved):
The change looks great @mayya-sharipova! I left some comments but nothing major.

jtibshirani reviewed (commented):
I noticed some small points while getting up to speed with the change !

mayya-sharipova committed:
Address Jim's feedback

mayya-sharipova commented on x-pack/plugin/mapper-unsigned-long/src/main/java/org/elasticsearch/xpack/unsignedlong/UnsignedLongFieldMapper.java:
```diff
@@ -0,0 +1,561 @@
+        }
+    }
+
+    public static final class UnsignedLongFieldType extends SimpleMappedFieldType {
```
@jtibshirani Thanks for the feedback, it did not make sense for me to extend `NumberFieldType` as logic for almost every query needs to be overwritten.

jtibshirani replied:
Thanks for the explanation.

mayya-sharipova committed:
Include unsigned_long docs into numeric type

mayya-sharipova commented on docs/reference/mapping/types.asciidoc:
```diff
@@ -9,7 +9,7 @@ document:
 
 string::         <<text,`text`>>, <<keyword,`keyword`>> and <<wildcard,`wildcard`>>
-<<number>>::     `long`, `integer`, `short`, `byte`, `double`, `float`, `half_float`, `scaled_float`
+<<number>>::     `long`, `integer`, `short`, `byte`, `double`, `float`, `half_float`, `scaled_float`, <<unsigned-long,`unsigned_long`>>
```
@jtibshirani Good suggestion, addressed in ab54a23

mayya-sharipova committed:
Merge remote-tracking branch 'upstream/master' into unsigned64bits_integer

mayya-sharipova committed:
Convert UnsignedLongFieldMapper to parametrized

- Convert UnsignedLongFieldMapper to a parametrized form
- Small adjustments in UnsignedLongScriptDocValues

jimczi commented on docs/reference/mapping/types/unsigned_long.asciidoc:
```diff
@@ -0,0 +1,115 @@
+if the value of this document is within the range of long values,
+or of the type `BigInteger` if the value exceeds this range.
+
+WARNING: Not all {es} clients can properly handle big integer values.
```
That's a bit scary to add as a warning ;). I would change {es} with something like `rest clients need to handle big integer values in json to support this field type correctly` or something along those lines ?

jimczi commented on docs/reference/mapping/types/unsigned_long.asciidoc:
```diff
@@ -0,0 +1,115 @@
+//TEST[continued]
+
+Similarly to sort values, script values of an `unsigned_long` field
+produce `BigInteger` or `Long` values. The same values: `BigInteger` or
```
I'd say `produce a `Number` that represents a long or a BigInteger...`

jimczi reviewed (approved):
This change looks good to me.

mayya-sharipova committed:
Small edits in documentation

mayya-sharipova commented:
@elasticmachine run elasticsearch-ci/2

mayya-sharipova commented:
@jtibshirani Thanks for the feedback so far, I was wondering if you had additional comments.

jtibshirani commented on docs/reference/mapping/types/unsigned_long.asciidoc:
```diff
@@ -0,0 +1,116 @@
+return a number representing a `long` or `BigInteger`.
+The same values: `long` or `BigInteger` are used for `terms` aggregations.
+
+==== Queries with mixed numeric types
```
Small comment, we could say 'Searches with mixed numeric types' to be more precise (since this section mentions aggs too).

mayya-sharipova replied:
Addressed in 17912bc

jtibshirani commented on docs/reference/mapping/types/unsigned_long.asciidoc:
```diff
@@ -0,0 +1,116 @@
+if the value of this document is within the range of long values,
+or of the type `BigInteger` if the value exceeds this range.
+
+NOTE: Rest clients need to be able to handle big integer values
```
Really small comment, I think we usually capitalize REST and JSON.

mayya-sharipova replied:
Addressed in 17912bc

jtibshirani commented on docs/reference/mapping/types/unsigned_long.asciidoc:
```diff
@@ -0,0 +1,116 @@
+--------------------------------
+//TEST[continued]
+
+Similarly to sort values, script values of an `unsigned_long` field
```
I think it'd be clearer to call out the Java type here: "return a `Number` representing a `Long` or `BigInteger`".

mayya-sharipova replied:
Addressed in 17912bc

jtibshirani reviewed (commented):
@mayya-sharipova thanks for the ping. I did a final review, just focusing on the docs/ high-level behavior. I had one remaining question: should we throw an error when attempting to sort across indices with mixed `long` and `unsigned_long`? I could see this behavior being really surprising and hard to track down for users.

mayya-sharipova committed:
Address Julie's comment on documentation

mayya-sharipova committed:
Add check that unsigned_long field type can't be sorted with other types

mayya-sharipova committed:
Merge remote-tracking branch 'upstream/master' into unsigned64bits_integer

mayya-sharipova committed:
Fix build and test failures

jtibshirani commented on server/src/main/java/org/elasticsearch/action/search/SearchPhaseController.java:
```diff
@@ -485,6 +486,36 @@ private static InternalAggregations reduceAggs(InternalAggregation.ReduceContext
+        boolean firstResult = true;
+        for (SearchPhaseResult entry : queryResults) {
+            DocValueFormat[] formats = entry.queryResult().sortValueFormats();
+            if (formats == null) return;
```
Checking my understanding -- are all shards guaranteed to have the same number of sort formats (even if some sort fields are unmapped)?

mayya-sharipova replied:
@jtibshirani Indeed, all shards are guaranteed to have the same number of sort formats. If a field is unmapped on a shard, we will get a shard failure, unless we specifically was [map an unmapped to something else](https://www.elastic.co/guide/en/elasticsearch/reference/current/sort-search-results.html#_ignoring_unmapped_fields).

mayya-sharipova replied:
Also wondering if @jimczi is ok with this final check I've added in `SearchPhaseController`

jimczi replied:
++, that makes sense to me. You can maybe rename the function validateMergeSortValueFormats or something along those lines ?

mayya-sharipova commented:
@jtibshirani Thanks for the feedback.

> should we throw an error when attempting to sort across indices with mixed long and unsigned_long?

I have added this check in b2eef4c

jtibshirani commented on server/src/main/java/org/elasticsearch/action/search/SearchPhaseController.java:
```diff
@@ -485,6 +486,36 @@ private static InternalAggregations reduceAggs(InternalAggregation.ReduceContext
+     * Checks that query results from all shards have consistent unsigned_long format.
+     * Sort queries on a field that has long type in one index, and unsigned_long in another index
+     * don't work correctly. Throw an error if this kind of sorting is detected.
+     * //TODO: instead of throwing error, find a way to sort long and unsigned_long together
```
This TODO makes sense. It's unfortunate that we need to do a special check here, but it feels worth it to me to avoid silently returning incorrect results.

jtibshirani reviewed (approved):
It looks good to me from the high-level behavior + docs side!

mayya-sharipova committed:
Merge remote-tracking branch 'upstream/master' into unsigned64bits_integer

mayya-sharipova committed:
Rename the method for validating consistency of merge formats

mayya-sharipova committed:
Change unsigned_long mapper based on recent master changes
